modules/stitching/src/matchers.cpp

```diff
@@ -583,7 +583,10 @@ void SurfFeaturesFinderGpu::find(InputArray image, ImageFeatures &features)
     image_.upload(image);
 
     ensureSizeIsEnough(image.size(), CV_8UC1, gray_image_);
-    cvtColor(image_, gray_image_, COLOR_BGR2GRAY);
+
+    Mat gray_image;
+    cvtColor(image, gray_image, COLOR_BGR2GRAY);
+    gray_image_.upload(gray_image);
```

Comment:
Though this solves the problem, thinking about it more now I feel that since surf explicitly needs a grayscale image, converting image to grayscale should not be the job of this function. The best it can do is to assert if the image is grayscale or not. So the change should be in other places which want to use surf matchers.